File: web/src/models/issue.ts
import { Policy } from "./policy"
import { Repository } from "./repository"

export interface Commit {
  author?: {
    date: Date,
    email: string,
    name: string,
    timezone: string
  },
  email: string,
  hash: string,
}

export interface Offender {
  object: string,
  operand: string,
  operator: string,
  value: string
}

export interface Issue {
  id: string,
  commit: Commit,
  condition_type: string,
  file?: string,
  offender?: Offender,
  policy?: Policy,
  repository?: Repository
  severity: number
}

File: internal/core/models/issue.go
package models

// Issue analysis issue
type Issue struct {
	ID            string        `json:"id,omitempty"`
	Commit        Commit        `json:"commit"`
	ConditionType ConditionType `json:"condition_type"`
	Leaks         []Leak        `json:"leaks,omitempty"`
	Message       string        `json:"message"`
	Offender      *Offender     `json:"offender,omitempty"`
	Policy        Policy        `json:"policy"`
	PolicyType    PolicyType    `json:"policy_type"`
	Repository    *Repository   `json:"repository,omitempty"`
	Severity      Score         `json:"severity"`
}

type Offender struct {
	Object   string `json:"object,omitempty"`
	Operator string `json:"operator,omitempty"`
	Operand  string `json:"operand,omitempty"`
	Value    string `json:"value,omitempty"`
}

// WithCommit
func (issue *Issue) WithCommit(commit Commit) {
	issue.Commit = commit
}

// WithLeak attach leaks to issue
func (issue *Issue) WithLeak(leak Leak) {
	issue.Leaks = append(issue.Leaks, leak)
}

// WithLeaks attach leaks to issue
func (issue *Issue) WithLeaks(leaks []Leak) {
	issue.Leaks = leaks
}

func NewIssue() *Issue {
	return &Issue{}
}
